tests/test_beamtime.py

```diff
@@ -0,0 +1,322 @@
+        self.assertEqual(bss_info.CONNECT_TIMEOUT, 5)
+        self.assertEqual(bss_info.POLL_INTERVAL, 0.01)
+        self.assertEqual(
+            bss_info.DM_APS_DB_WEB_SERVICE_URL,
+            "https://xraydtn01.xray.aps.anl.gov:11236")
+        self.assertIsNotNone(bss_info.api_bss)
+        self.assertIsNotNone(bss_info.api_esaf)
+
+    def test_iso2datetime(self):
+        self.assertEqual(
+            bss_info.iso2datetime("2020-06-30 12:31:45.067890"),
+            datetime.datetime(2020, 6, 30, 12, 31, 45, 67890)
+        )
+
+    def test_not_at_aps(self):
+        self.assertTrue(True)   # test something
+        if using_APS_workstation():
+            return
+
+        # do not try to test for fails using dm package, it has no timeout
+
+    def test_only_at_aps(self):
+        self.assertTrue(True)   # test something
+        if not using_APS_workstation():
+            return
+
+        runs = bss_info.listAllRuns()
+        self.assertGreater(len(runs), 1)
+        self.assertEqual(bss_info.getCurrentCycle(), runs[-1])
+        self.assertEqual(bss_info.listRecentRuns()[0], runs[-1])
+
+        self.assertGreater(len(bss_info.listAllBeamlines()), 1)
+
+        # TODO: test the other functions
+        # getCurrentEsafs
+        # getCurrentInfo
+        # getCurrentProposals
+        # getEsaf
+        # getProposal
+        # class DmRecordNotFound(Exception): ...
+        # class EsafNotFound(DmRecordNotFound): ...
+        # class ProposalNotFound(DmRecordNotFound): ...
+        # EsafInfo
+        # ScheduleInfo
+
+    def test_printColumns(self):
+        from tests.common import Capture_stdout
+        with Capture_stdout() as received:
+            bss_info.printColumns("1 2 3 4 5 6".split(), numColumns=3, width=3)
+        self.assertEqual(len(received), 2)
+        self.assertEqual(received[0], "1  3  5  ")
+        self.assertEqual(received[1], "2  4  6  ")
+
+        source = "0123456789"
+        self.assertEqual(bss_info.trim(source), source)
+        self.assertNotEqual(bss_info.trim(source, length=8), source)
+        self.assertEqual(bss_info.trim(source, length=8), "01234...")
+
+
+class Test_EPICS(unittest.TestCase):
+
+    def start_ioc_subprocess(self, db_file):
+        softIoc = os.path.abspath(os.path.join(
+            os.environ.get("CONDA_PREFIX"),
+            "epics",
+            "bin",
+            os.environ.get("EPICS_HOST_ARCH", "linux-x86_64"),
+            "softIoc"
+        ))
+        if not os.path.exists(softIoc):
+            return
+        db_file = os.path.abspath(os.path.join(
+            os.path.dirname(bss_info.__file__),
+            db_file
+        ))
+        if not os.path.exists(db_file):
+            return
+        cmd = f"{softIoc} -m P={BSS_TEST_IOC_PREFIX} -d {db_file}".split()
+        return subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=False)
```

Comment:
Codacy found an issue: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/gh/BCDA-APS/apstools/pullRequest?prid=5764884)